include/eve/traits/overload/default_behaviors.hpp

```diff
@@ -0,0 +1,157 @@
+//======================================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//======================================================================================================================
+#pragma once
+
+#include <eve/concept/value.hpp>
+#include <eve/detail/skeleton.hpp>
+#include <eve/detail/function/conditional.hpp>
+
+namespace eve
+{
+  namespace detail
+  {
+    // Usual object function so we can have EVE_FORCEINLINE there for QoI on ARM/AARCH64
+    inline struct { EVE_FORCEINLINE auto operator()(auto, auto x) const { return x; } } eval_2nd = {};
+  }
+
+  //====================================================================================================================
+
+  //====================================================================================================================
+  template< template<typename> class Func
+          , typename OptionsValues
+          , typename... Options
+          >
+  struct callable : decorated_with<OptionsValues, Options...>
+  {
+    using base = decorated_with<OptionsValues, Options...>;
+
+    template<typename T>
+    auto operator[](T t) const
+    requires( requires(base const& b) { b[t];} )
+    {
+      auto new_traits = base::operator[](t);
+      return  Func<decltype(new_traits)>{new_traits};
+    }
+
+    template<typename T> void operator[](T t) const =delete;
```

Comment:
Can you add a comment that this only helps with error messages.
Also - does it? I remeber we tried but I don't remember the results.

Reply:
# With =delete
## g++
```c++
<source>: In function ‘int main()’:
<source>:42:33: error: use of deleted function ‘constexpr void eve::callable<Func, OptionsValues, Options>::operator[](T) const [with T = const char*; Func = eve::func_t; OptionsValues = eve::options<>; Options = {eve::conditional_option}]’
   42 |   std::cout << eve::func["false"](8) << "\n";
      |                                 ^
In file included from /home/falcou/dev/eve/include/eve/traits/overload.hpp:12,
                 from <source>:3:
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:41:41: note: declared here
   41 |     template<typename T> constexpr void operator[](T t) const =delete;
      |                                         ^~~~~~~~
```

## clang
```c++
<source>:42:25: error: overload resolution selected deleted operator '[]'
  std::cout << eve::func["false"](8) << "\n";
               ~~~~~~~~~^~~~~~~~
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:41:41: note: candidate function [with T = const char *] has been explicitly deleted
    template<typename T> constexpr void operator[](T t) const =delete;
                                        ^
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:34:20: note: candidate template ignored: constraints not satisfied [with T = const char *]
    constexpr auto operator[](T t) const
                   ^
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:35:41: note: because 'b[t]' would be invalid: no viable overloaded operator[] for type 'const eve::callable<eve::func_t, eve::options<>, eve::conditional_option>::base' (aka 'const decorated_with<eve::options<rbr::settings<>>, eve::conditional_option>')
    requires( requires(base const& b) { b[t];} )
                                        ^
```

#  Without `=delete`

## g++
```c++
<source>: In function ‘int main()’:
<source>:42:25: error: no match for ‘operator[]’ (operand types are ‘const eve::func_t<eve::options<> >’ and ‘const char [6]’)
   42 |   std::cout << eve::func["false"](8) << "\n";
      |                         ^
In file included from /home/falcou/dev/eve/include/eve/traits/overload.hpp:12,
                 from <source>:3:
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:34:20: note: candidate: ‘template<class T> constexpr auto eve::callable<Func, OptionsValues, Options>::operator[](T) const requires requires(const eve::base& b) {(b)[eve::callable<Func, OptionsValues, Options ...>::operator[]::t];} [with Func = eve::func_t; OptionsValues = eve::options<>; Options = {eve::conditional_option}]’
   34 |     constexpr auto operator[](T t) const
      |                    ^~~~~~~~
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:34:20: note:   template argument deduction/substitution failed:
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:34:20: note: constraints not satisfied
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp: In substitution of ‘template<class T> constexpr auto eve::callable<eve::func_t, eve::options<>, eve::conditional_option>::operator[](T) const requires requires(const eve::base& b) {(b)[eve::callable<Func, OptionsValues, Options ...>::operator[]::t];} [with T = eve::func_t]’:
<source>:42:33:   required from here
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:34:20:   required by the constraints of ‘template<template<class> class Func, class OptionsValues, class ... Options> template<class T> constexpr auto eve::callable<Func, OptionsValues, Options>::operator[](T) const requires requires(const eve::base& b) {(b)[eve::callable<Func, OptionsValues, Options ...>::operator[]::t];}’
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:35:15:   in requirements with ‘const eve::callable<Func, OptionsValues, Options>::base& b’ [with T = const char*; Func = eve::func_t; OptionsValues = eve::options<rbr::settings<> >; Options = {eve::conditional_option}]
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:35:42: note: the required expression ‘b[t]’ is invalid
   35 |     requires( requires(base const& b) { b[t];} )
      |                                         ~^
cc1plus: note: set ‘-fconcepts-diagnostics-depth=’ to at least 2 for more detail
```

## clang
```c++
<source>:42:25: error: no viable overloaded operator[] for type 'const func_t<eve::options<>>'
  std::cout << eve::func["false"](8) << "\n";
               ~~~~~~~~~^~~~~~~~
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:34:20: note: candidate template ignored: constraints not satisfied [with T = const char *]
    constexpr auto operator[](T t) const
                   ^
/home/falcou/dev/eve/include/eve/traits/overload/default_behaviors.hpp:35:41: note: because 'b[t]' would be invalid: no viable overloaded operator[] for type 'const eve::callable<eve::func_t, eve::options<>, eve::conditional_option>::base' (aka 'const decorated_with<eve::options<rbr::settings<>>, eve::conditional_option>')
    requires( requires(base const& b) { b[t];} )
```

# Conclusion
 - For clang, it is about the same
  - For g++, it prevents this long cascade of "or else..."

Reply:
add a comment anyway please